File: worker/src/lib/evidenceContracts.ts
import contractManifest from '../../../schemas/expected-return-contracts-v1.json'

type ManifestContract = {
  artifact_contract_version: string
  feature_semantic_version: string
  label_schema_version: string
  model_version_prefix: string
  validation_schema_version: string
  expected_return_semantic: string
}

const manifest = contractManifest as {
  schema_version: string
  canonical_label_schema_version: string
  legacy_label_schema_version: string
  canonical_roundtrip_cost_bps: number
  owners: {
    l4_alpha_ev: { current: ManifestContract; retired_compatibility: Record<string, unknown>[] }
    allocator_ev_fusion: { current: ManifestContract; retired_compatibility: Record<string, unknown>[] }
  }
}

function runtimeContract(value: ManifestContract) {
  return Object.freeze({
    artifactContractVersion: value.artifact_contract_version,
    featureSemanticVersion: value.feature_semantic_version,
    labelSchemaVersion: value.label_schema_version,
    modelVersionPrefix: value.model_version_prefix,
    validationSchemaVersion: value.validation_schema_version,
    expectedReturnSemantic: value.expected_return_semantic,
  })
}

export const EXPECTED_RETURN_CONTRACT_MANIFEST_VERSION = manifest.schema_version
export const EVIDENCE_LABEL_SCHEMA_VERSION = manifest.canonical_label_schema_version
export const LEGACY_EVIDENCE_LABEL_SCHEMA_VERSION = manifest.legacy_label_schema_version
export const CANONICAL_ROUNDTRIP_COST_BPS = manifest.canonical_roundtrip_cost_bps
export const RETIRED_EXPECTED_RETURN_COMPATIBILITY = Object.freeze({
  l4_alpha_ev: Object.freeze([...manifest.owners.l4_alpha_ev.retired_compatibility]),
  allocator_ev_fusion: Object.freeze([...manifest.owners.allocator_ev_fusion.retired_compatibility]),
})

export const L4_ALPHA_EV_CONTRACT = runtimeContract(manifest.owners.l4_alpha_ev.current)

const fusionCurrent = runtimeContract(manifest.owners.allocator_ev_fusion.current)
export const ALLOCATOR_EV_FUSION_CONTRACT = Object.freeze({
  ...fusionCurrent,
  compatiblePairs: Object.freeze([fusionCurrent]),
})

File: ml-controller/services/evidence_contracts.py
"""Canonical cross-layer versions loaded from the repository contract manifest."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


def _manifest_path() -> Path:
    source = Path(__file__).resolve()
    candidates = (
        source.parents[2] / "schemas" / "expected-return-contracts-v1.json",
        source.parents[1] / "schemas" / "expected-return-contracts-v1.json",
    )
    for candidate in candidates:
        if candidate.exists():
            return candidate
    raise RuntimeError("expected_return_contract_manifest_missing")


EXPECTED_RETURN_CONTRACT_MANIFEST_PATH = _manifest_path()
EXPECTED_RETURN_CONTRACT_MANIFEST: dict[str, Any] = json.loads(
    EXPECTED_RETURN_CONTRACT_MANIFEST_PATH.read_text(encoding="utf-8")
)
if EXPECTED_RETURN_CONTRACT_MANIFEST.get("schema_version") != "expected-return-contract-manifest-v1":
    raise RuntimeError("expected_return_contract_manifest_version_invalid")

LABEL_SCHEMA_VERSION = str(EXPECTED_RETURN_CONTRACT_MANIFEST["canonical_label_schema_version"])
LEGACY_LABEL_SCHEMA_VERSION = str(EXPECTED_RETURN_CONTRACT_MANIFEST["legacy_label_schema_version"])
CANONICAL_ROUNDTRIP_COST_BPS = float(
    EXPECTED_RETURN_CONTRACT_MANIFEST["canonical_roundtrip_cost_bps"]
)
CANONICAL_ROUNDTRIP_COST_RATE = CANONICAL_ROUNDTRIP_COST_BPS / 10_000.0

_L4 = EXPECTED_RETURN_CONTRACT_MANIFEST["owners"]["l4_alpha_ev"]
_FUSION = EXPECTED_RETURN_CONTRACT_MANIFEST["owners"]["allocator_ev_fusion"]
_L4_CURRENT = _L4["current"]
_FUSION_CURRENT = _FUSION["current"]

L4_ARTIFACT_CONTRACT_VERSION = str(_L4_CURRENT["artifact_contract_version"])
L4_FEATURE_SEMANTIC_VERSION = str(_L4_CURRENT["feature_semantic_version"])
L4_EXPECTED_RETURN_SEMANTIC = str(_L4_CURRENT["expected_return_semantic"])
SUPPORTED_L4_SERVING_CONTRACT_PAIRS = frozenset({
    (L4_ARTIFACT_CONTRACT_VERSION, LABEL_SCHEMA_VERSION),
})
SUPPORTED_L4_FEATURE_SEMANTICS = {
    L4_ARTIFACT_CONTRACT_VERSION: L4_FEATURE_SEMANTIC_VERSION,
}
RETIRED_L4_COMPATIBILITY = tuple(_L4.get("retired_compatibility") or ())

ALLOCATOR_EV_ARTIFACT_CONTRACT_VERSION = str(_FUSION_CURRENT["artifact_contract_version"])
ALLOCATOR_EV_FEATURE_SEMANTIC_VERSION = str(_FUSION_CURRENT["feature_semantic_version"])
ALLOCATOR_EV_EXPECTED_RETURN_SEMANTIC = str(_FUSION_CURRENT["expected_return_semantic"])
SUPPORTED_ALLOCATOR_EV_SERVING_CONTRACT_PAIRS = frozenset({
    (ALLOCATOR_EV_ARTIFACT_CONTRACT_VERSION, LABEL_SCHEMA_VERSION),
})
SUPPORTED_ALLOCATOR_EV_FEATURE_SEMANTICS = {
    ALLOCATOR_EV_ARTIFACT_CONTRACT_VERSION: ALLOCATOR_EV_FEATURE_SEMANTIC_VERSION,
}
RETIRED_ALLOCATOR_EV_COMPATIBILITY = tuple(_FUSION.get("retired_compatibility") or ())
